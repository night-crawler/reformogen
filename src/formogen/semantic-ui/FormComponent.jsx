import React from 'react';
import PropTypes from 'prop-types';
import { Button, Header, Form as SUIForm, Grid } from 'semantic-ui-react';
import { isEmpty } from 'lodash';

import { NonFieldErrorsList } from './NonFieldErrorsList';

function DefaultSubmitComponent(props) {
  const { style, ...rest } = props;
  const patchedStyle = {
    marginTop: '20px',
    ...style,
  };
  return <Button style={ patchedStyle } { ...rest } />;
}


FormComponent.displayName = 'FormComponent';
FormComponent.propTypes = {
  loading: PropTypes.bool.isRequired,

  title: PropTypes.string,
  isTitleVisible: PropTypes.bool,

  nonFieldErrorsMap: PropTypes.object,
  formLayout: PropTypes.array.isRequired,

  submitComponent: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.element,
    PropTypes.instanceOf(React.Component)
  ]),
  onSubmit: PropTypes.func,
};
FormComponent.defaultProps = {
  loading: false,
  title: 'Formogen Form',
  isTitleVisible: true,
  submitComponent: DefaultSubmitComponent,
};
export function FormComponent(props) {
  const { submitComponent: SubmitComponent } = props;
  return (
    <SUIForm loading={ props.loading }>
      { props.isTitleVisible
        ? <Header as='h2' dividing={ true }>{ props.title }</Header>
        : null
      }
      <div className='layouts'>
        { !isEmpty(props.nonFieldErrorsMap) &&
          <NonFieldErrorsList errors={ props.nonFieldErrorsMap } />
        }
        { props.formLayout.map(({ header, fields }, i) =>
          <Grid key={ `layout:${i}` } columns={ 16 } className='layout'>
            { !!header &&
              <div className='sixteen wide column'>
                <Header>{ header }</Header>
              </div>
            }
            { fields }
          </Grid>
        ) }
      </div>
      <SubmitComponent
        type='submit'
        content='Submit'
        fluid={ true }

        onClick={ props.onSubmit }
      />
    </SUIForm>
  );
}
