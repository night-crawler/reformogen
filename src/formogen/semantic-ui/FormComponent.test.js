import { FormComponent } from './FormComponent';


it('<FormComponent />', () => {
  const wrapper = shallow(
    <FormComponent
      loading={ true }
      title={ 'sample' }
      isTitleVisible={ true }
      fieldsets={ [] }
    />
  );
  expect(wrapper).toMatchSnapshot();
});
