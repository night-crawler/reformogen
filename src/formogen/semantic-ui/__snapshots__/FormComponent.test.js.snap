// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormComponent /> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormComponent
    fieldsets={Array []}
    isTitleVisible={true}
    loading={true}
    onSubmit={[Function]}
    submitComponent={[Function]}
    title="sample"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "as": "form",
      "children": Array [
        <Header
          as="h2"
          dividing={true}
        >
          sample
        </Header>,
        <div
          className="layouts"
        />,
        <DefaultSubmitComponent
          content="Submit"
          fluid={true}
          onSubmit={[Function]}
          type="submit"
        />,
      ],
      "loading": true,
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "h2",
          "children": "sample",
          "dividing": true,
        },
        "ref": null,
        "rendered": "sample",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            false,
            Array [],
          ],
          "className": "layouts",
        },
        "ref": null,
        "rendered": Array [
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "content": "Submit",
          "fluid": true,
          "onSubmit": [Function],
          "type": "submit",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "as": "form",
        "children": Array [
          <Header
            as="h2"
            dividing={true}
          >
            sample
          </Header>,
          <div
            className="layouts"
          />,
          <DefaultSubmitComponent
            content="Submit"
            fluid={true}
            onSubmit={[Function]}
            type="submit"
          />,
        ],
        "loading": true,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "h2",
            "children": "sample",
            "dividing": true,
          },
          "ref": null,
          "rendered": "sample",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              false,
              Array [],
            ],
            "className": "layouts",
          },
          "ref": null,
          "rendered": Array [
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "content": "Submit",
            "fluid": true,
            "onSubmit": [Function],
            "type": "submit",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
